src/clinical/clinical-service.ts

```diff
@@ -355,60 +386,76 @@ export const getValidRecordsPostSubmission = async (
 
   const validDonorIds = donorData
     .filter(donor => donor.schemaMetadata.isValid)
-    .map(({ donorId }) => donorId);
-
-  const clinicalErrors: ClinicalErrorsResponseRecord[] = [];
+    .map(({ donorId }) => donorId)
+    .filter(id => typeof id === 'number');
 
   const invalidDonorIds = errorDonorIds.filter(
     (donorId, index, array) =>
       !validDonorIds.includes(donorId) && filterDuplicates(donorId, index, array),
   );
 
-  for (const donorId of invalidDonorIds) {
-    const currentDonor = donorData.filter(donor => donor.donorId === donorId)[0];
-    const { submitterId: submitterDonorId } = currentDonor;
-    const currentDonorErrors = clinicalMigrationErrors
-      .filter(errorRecord => errorRecord.donorId === donorId)
-      .map(migrationError => migrationError.errors)
-      .flat();
-
-    const currentDonorEntities = currentDonorErrors
-      .map(error => error.entityName)
-      .filter(filterDuplicates);
+  let clinicalErrors: ClinicalErrorsResponseRecord[] = [];
 
+  if (invalidDonorIds.length > 0) {
     const migrationVersion =
       lastMigration?.toVersion || (await dictionaryManager.instance().getCurrentVersion());
+
     const schemaName = await dictionaryManager.instance().getCurrentName();
 
     const migrationDictionary = await dictionaryManager
       .instance()
       .loadSchemaByVersion(schemaName, migrationVersion);
 
-    currentDonorEntities.forEach(entityName => {
-      const entityValidationErrors =
-        MigrationManager.validateDonorEntityAgainstNewSchema(
-          entityName,
-          migrationDictionary,
-          currentDonor,
-        ) || [];
-
-      const entityErrorRecords: ClinicalEntityErrorRecord[] = entityValidationErrors.map(
-        (validationRecord): ClinicalEntityErrorRecord => ({
-          donorId,
-          entityName,
-          ...validationRecord,
-        }),
-      );
-
-      const errorResponseRecord: ClinicalErrorsResponseRecord = {
-        donorId,
-        submitterDonorId,
-        entityName,
-        errors: entityErrorRecords,
-      };
-
-      if (entityErrorRecords.length) clinicalErrors.push(errorResponseRecord);
-    });
+    const invalidDonorRecords = donorData.filter(
+      donor => !!donor.donorId && invalidDonorIds.includes(donor.donorId),
+    ) as DeepReadonly<DonorWithId>[];
+
+    const clinicalErrorRecords = await Promise.all(
```

Comment:
a comment would be helpful here. there is a lot to follow with multiple Promise arrays and nested mapping and then just returning the 1st index place of the array.

Reply:
Good call, will add

Reply:
Added